### ISSUE TITLE
Your mono runtime and class libraries are out of sync. mscorlib.dll

### ISSUE THREAD

**Issue by Trider12** [godotengine/godot#45550]
**Godot version:**
3.2.3.stable.mono

**OS/device including version:**
Windows 10

**Issue description:**
I compiled Godot using Mono 6.12.0 Stable and this [tutorial](https://docs.godotengine.org/en/stable/development/compiling/compiling_with_mono.html) and then created export templates according to [this](https://docs.godotengine.org/en/stable/development/compiling/compiling_for_windows.html#creating-windows-export-templates) one. My game worked fine in the editor, but after exporting it, I encountered this error launching the executable:
![image](https://user-images.githubusercontent.com/47952308/106231318-9ba73e00-621b-11eb-9a49-22b7e0850fda.png)
After copying `mono-2.0-sgen.dll` from Godot `bin` folder to the `.exe` folder of my game I see this message in the debug console:
```
Your mono runtime and class libraries are out of sync.
The out of sync library is: .mono/assemblies/Debug/mscorlib.dll

When you update one from git you need to update, compile and install
the other too.
Do not report this as a bug unless you're sure you have updated correctly:
you probably have a broken mono install.
If you see other errors or faults after this message they are probably related
and you need to fix your mono install first.

Your mono runtime and class libraries are out of sync.
The out of sync library is: .mono/assemblies/Debug/mscorlib.dll

When you update one from git you need to update, compile and install
the other too.
Do not report this as a bug unless you're sure you have updated correctly:
you probably have a broken mono install.
If you see other errors or faults after this message they are probably related
and you need to fix your mono install first.
```
But when I check the `.mono/assemblies/Debug/` directory of the project, there is no `mscorlib.dll` there.

**Comment by thomassmith311** [comment]
Does the file mscorlib.dll appear when you view hidden items?
![image](https://user-images.githubusercontent.com/19561856/106358565-08cdd880-62db-11eb-950c-998e66bd4032.png)

**Comment by Trider12** [comment]
@thomassmith311, no, it does not (is this a genuine question?)

**Comment by aaronfranke** [comment]
@Trider12 Do you have both Mono and .NET Core SDK installed, and are they 64-bit?

**Comment by Trider12** [comment]
@aaronfranke, the Mono install is 64-bit and the .NET SDK's are the ones shipped with VS2019 (both 32 and 64-bit).

**Comment by Trider12** [comment]
@akien-mga, I'm sorry, but why is this question labeled as `documentation`? Is there a not yet documented workaround to this issue?

**Comment by Trider12** [comment]
I've got rid of the `mono-2.0-sgen.dll` error by compiling with `mono_static=yes`. This is, indeed, [documented](https://docs.godotengine.org/en/stable/development/compiling/compiling_with_mono.html#rebuild-with-mono-glue) (my bad), but I find the line `Make sure to include this library when distributing Godot.` a bit misleading, as I've previously thought it to be referring to the engine itself and not the exported game.
Nevertheless, the second issue still persists.

**Comment by jjmontesl** [comment]
I'm facing the same issue with exported templates, but I'm running on Ubuntu 18.04.

I can see `.mono/assemblies/Release/mscorlib.dll` inside the `.pck`, if that matters.

**Comment by jjmontesl** [comment]
It seems to me that Godot is including `mscorlib.dll` always from ~/.local/share/godot/templates/3.2.3.stable.mono/bcl/net_4_x/  or (net_4_x_win).

I am trying by compiling Mono and generating its corresponding bcl directory (following https://github.com/godotengine/godot-mono-builds), and when I copy the resulting bcl directories to the default template seems to work 

It seems that specifying a custom template does not make Godot pick the BCL directory from that directory (and atm I am unsure where the data directory is copied from).

Side question: Is there any way of selecting a different "default" template, seems it's picked by name automatically?

**Comment by Calinou** [comment]
> Side question: Is there any way of selecting a different "default" template, seems it's picked by name automatically?

The default template is selected depending on the Godot editor version.

If you want to use different data path directories for a specific Godot editor build you're using, you can enable [self-contained mode](https://docs.godotengine.org/en/latest/tutorials/io/data_paths.html#self-contained-mode).

**Comment by akien-mga** [comment]
> It seems that specifying a custom template does not make Godot pick the BCL directory from that directory (and atm I am unsure where the data directory is copied from).

Yes I don't think that the custom template option in the export preset supports locating and using a custom BCL. This option exists for the classical build which only need a self-contained binary, but exporting for Mono is more complex and I doubt @neikeq built in the necessary flexibility.

So you'll have to add your own templates and BCL to `~/.local/share/godot/templates/3.2.3.stable.mono` to replace the upstream ones, or do it with the self-contained mode as suggested by @Calinou.

**Comment by jjmontesl** [comment]
One side note in case someone else hits this: it seems that the official template builds are generated by the scripts at https://github.com/godotengine/build-containers .  I'm trying to use those as a reference.

(This is needed for bugfixing, script encryption, adding modules... and it's always good to try and release something as close as what users had, so it's interesting to know how official templates were compiled, optimized, etc).

This is (afaik) not mentioned in the build / export templates documentation. It might be interesting to state that the build-containers are the official way of building the editor and templates, in order to foster reproducible builds.

**Comment by akien-mga** [comment]
>  (EDIT) Some build scripts are still missing in that repo, and it'd be ideal if one could just pass in a godot repo path and it'd build the templates and tools

The build scripts are in https://github.com/godotengine/godot-build-scripts

Neither of those two repos are officially supported for end users though, it's what I use to make official builds and it *can* be used to replicate that, but it's not user-friendly in any way. We're working on a more user-friendly solution for user custom builds, but it's not ready yet.

**Comment by Trider12** [comment]
I revisited the issue with `godot-3.3.1-stable` and `Mono 6.12.0 Stable`. 
I compiled the engine, enabled self-contained mode and replaced the default templates with the compiled ones.

Everything seems to be working fine now.

**Comment by neikeq** [comment]
Regarding build scripts for end user custom builds, this likely won't be needed starting with Godot 4.1 and later as Godot won't be linking Mono nor any other .NET runtime anymore. A priori, the only build step that will remain is building the C# API and editor tools C# code.

**Comment by akien-mga** [comment]
Is this still reproducible in Godot 3.6 or later?

**Comment by Calinou** [comment]
Closing due to lack of response. Please comment if you can still reproduce this bug on the latest Godot version.
